pkg/agent/sysadvisor/plugin/qosaware/resource/memory/plugin/memory_offloading.go

```diff
@@ -0,0 +1,543 @@
+	"github.com/kubewharf/katalyst-core/pkg/agent/sysadvisor/types"
+	"github.com/kubewharf/katalyst-core/pkg/config"
+	"github.com/kubewharf/katalyst-core/pkg/config/agent/dynamic/tmo"
+	tmoconf "github.com/kubewharf/katalyst-core/pkg/config/agent/dynamic/tmo"
+	"github.com/kubewharf/katalyst-core/pkg/consts"
+	katalystcoreconsts "github.com/kubewharf/katalyst-core/pkg/consts"
+	"github.com/kubewharf/katalyst-core/pkg/metaserver"
+	"github.com/kubewharf/katalyst-core/pkg/metrics"
+	"github.com/kubewharf/katalyst-core/pkg/util/cgroup/common"
+	"github.com/kubewharf/katalyst-core/pkg/util/general"
+	"github.com/kubewharf/katalyst-core/pkg/util/native"
+)
+
+const (
+	TransparentMemoryOffloading = "transparent-memory-offloading"
+)
+
+const (
+	InactiveProbe            = 0.1
+	OffloadingSizeScaleCoeff = 1.05
+)
+
+const (
+	DummyTMOBlockFnName string = "dummy-tmo-block-fn"
+)
+
+// TMO policy funcs to calculate memory offloading size
+var tmoPolicyFuncs sync.Map
+
+// TMO block funcs to filter the containers required to disable TMO
+var tmoBlockFuncs sync.Map
+
+type TmoStats struct {
+	memUsage                 float64
+	memInactive              float64
+	memPsiAvg60              float64
+	pgscan                   float64
+	pgsteal                  float64
+	refault                  float64
+	refaultActivate          float64
+	lastOffloadingTargetSize float64
+}
+
+type TmoPolicyFn func(
+	lastStats TmoStats,
+	currStats TmoStats,
+	conf *tmoconf.TMOConfigDetail) (error, float64)
+
+func psiPolicyFunc(lastStats TmoStats, currStats TmoStats, conf *tmoconf.TMOConfigDetail) (error, float64) {
+	if conf.PSIPolicyConf == nil {
+		return errors.New("psi policy requires psi policy configuration"), 0
+	}
+	return nil, math.Max(0, 1-(currStats.memPsiAvg60)/(conf.PSIPolicyConf.PsiAvg60Threshold)) * conf.PSIPolicyConf.MaxProbe * currStats.memUsage
+}
+
+func refaultPolicyFunc(lastStats TmoStats, currStats TmoStats, conf *tmoconf.TMOConfigDetail) (error, float64) {
+	if conf.RefaultPolicyConf == nil {
+		return errors.New("refault policy requires refault policy configurations"), 0
+	}
+	pgscanDelta := currStats.pgscan - lastStats.pgscan
+	pgstealDelta := currStats.pgsteal - lastStats.pgsteal
+	refaultDelta := currStats.refaultActivate - lastStats.refaultActivate
+	reclaimAccuracyRatio := 1.0
+	reclaimScanEfficiencyRatio := 1.0
+	if pgstealDelta > 0 && pgscanDelta > 0 {
+		reclaimAccuracyRatio = 1 - refaultDelta/pgstealDelta
+		reclaimScanEfficiencyRatio = pgstealDelta / pgscanDelta
+	}
+
+	if reclaimAccuracyRatio < conf.RefaultPolicyConf.ReclaimAccuracyTarget || reclaimScanEfficiencyRatio < conf.RefaultPolicyConf.ReclaimScanEfficiencyTarget {
+		// Decrease offloading size if detecting the reclaim accuracy or scan efficiency is below the targets
+		return nil, math.Max(0, currStats.lastOffloadingTargetSize*reclaimAccuracyRatio)
+	} else {
+		// Try to increase offloading size but make sure not exceed the max probe of memory usage and 10% of inactive memory
```

Comment:
it seems that it may exceed 10% of inactive memory because math.Max(currStats.lastOffloadingTargetSize*OffloadingSizeScaleCoeff, currStats.memInactive*InactiveProbe)